crates/geoip/src/resolver.rs

```diff
@@ -0,0 +1,29 @@
+use std::{net::IpAddr, sync::Arc};
+
+pub use {
+    crate::maxminddb::geoip2::City,
+    local::LocalResolver,
+    maxmind::{MaxMindResolver, MaxMindResolverError},
+};
+
+mod local;
+mod maxmind;
+
+#[derive(Debug, Clone)]
+pub struct GeoData {
+    pub continent: Option<Arc<str>>,
+    pub country: Option<Arc<str>>,
+    pub region: Option<Vec<String>>,
+    pub city: Option<Arc<str>>,
+}
+
+pub trait GeoIpResolver: Clone {
+    /// The error type produced by the resolver.
+    type Error;
+
+    /// Lookup the raw geo data for the given IP address.
+    fn lookup_geo_data_raw(&self, addr: IpAddr) -> Result<City<'_>, Self::Error>;
```

Comment:
Can we remove the MaxMind dependency from the trait? `lookup_geo_data_raw` can still be in the implementation but I don't see why it's needed here.

Reply:
It's needed as a low level optimization to not have to clone strings only to compare them. I don't see a problem in providing raw data for low level use.

Reply:
Oh, I misunderstood your comment. You're right that maxmind data should not be a part of generic trait. But the bigger problem here is the `LocalResolver` and the trait requirement. Do we really need the trait and the `LocalResolver`? It's a mock which is only used for tests, so we can try to come up with a different solution for testing.

Reply:
Thought about it, and I don't mind the dependency so much. `LocalResolver` is there to avoid the need to have a MaxMind DB around for testing, so as long as we can do that, I'm good.